### ISSUE TITLE
Some containers are not being monitored continuously and dissappear from the graphs (For example "OpenProject")

### ISSUE THREAD

**Issue by ektorasdj** [henrygd/beszel#58]
Hello,

Thank you for this nice tool. Unfortunately, I have some containers (created through Portainer) that are not continuously showing in the graphs. For example, check OpenProject, which is continuously running with no issues, but it disappears from the graphs for some time.
![image](https://github.com/user-attachments/assets/b0bcae2d-79dd-46d8-b06b-eaa566d7a164)

**Comment by henrygd** [comment]
Please see #50

It seems to be fixed by upgrading to the latest version of docker, or possibly just restarting the docker service. `sudo systemctl restart docker` if you're on Linux.

I'd love to get to the bottom of this and find the cause if it's fixable on this end. It would be helpful to know:

1. What version of docker are you on currently?
2. Has docker been running continuously for weeks without a restart?

**Comment by ektorasdj** [comment]
Hello, thank you for such a quick answer. Unfortunately i cannot update since its a synology NAS and custom docker updates bring a lot of other issues.

Here are my answers:

1. Version : 24.0.2 (currently the latest beta available version on Synology)

2.Last restart was 5 days ago.

In any case, its good to know that the latest update of docker fixes it since at some point in the future it will be fixed on my NAS.

**Comment by henrygd** [comment]
It looks like it could be a problem with the Docker Engine API in version 24.

Not all machines on 24 have the issue though, so it's possible that a restart could help. At least for a bit.

If someone comes along with the same problem on version 27 then I'll put it to the top of my list. But right now it's low priority since it seems to be fixable with an upgrade.

**Comment by henrygd** [comment]
Let me know if you can test this again. I'm curious if any of the updates we've made will fix this.

Also please try upgrading the Synology Container Manager if they've released an update.

Thanks!

**Comment by ektorasdj** [comment]
Hello!
I have tested this and while the issue is still here , it seems that the gaps are less frequent in the graphs.
![Screenshot_20240925-013042](https://github.com/user-attachments/assets/a1735a60-52ca-4c82-ae9d-e15d98dc9158)

Unfortunately Synology still uses the old docker package version 24 with no plans to update it as of yet.

Thank you!

**Comment by nathang21** [comment]
This is happening to me still, I have lots of gaps in both of the docker graphs.

I updated to the Beta Synology Container manager which is version, but the problem still occurs.

```
docker version
Client:
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        610b8d0
 Built:             Thu Aug  1 07:07:08 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       b5710a2
  Built:            Thu Aug  1 07:07:31 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.1
  GitCommit:        067f5021280b8de2059026fb5c43c4adb0f3f244
 runc:
  Version:          v1.1.7
  GitCommit:        adc1b13
 docker-init:
  Version:          0.19.0
  GitCommit:        ed96d00
```

**Comment by henrygd** [comment]
It's an issue with API version 1.43 and older. I think I figured out a workaround and will try to release it tomorrow.

**Comment by henrygd** [comment]
Let me know if 0.5.1 fixes this.

I can't test with Synology, but I was able to replicate and fix the issue using an LXC container running Docker 24.0.2.

**Comment by nathang21** [comment]
Hmmm unfortunately not, this is what my graphs look like this morning after I updated last night. 
<img width="1406" alt="image" src="https://github.com/user-attachments/assets/8826cfc6-c923-4781-8190-faba826cf153">
<img width="1472" alt="image" src="https://github.com/user-attachments/assets/024e5aa1-e29c-4d38-b63a-61f3f9d8720b">

For what it's worth, when I first installed Beszel I don't remember having this problem, and I was a pretty early adopter after seeing a post on r/selfhosted. I have added a lot of containers since then, so it could be my docker engine is just more overloaded now but figured i'd mention it. Happy to go back and test an older version if you have any suggestions.

**Comment by ektorasdj** [comment]
Hmm from my side it seems fixed (Synology , container manager-docker version 24.0.2 running all day after update).
![Screenshot_2](https://github.com/user-attachments/assets/a33727d4-5a7f-42bb-bf0f-fecb3d4451e1)

**Comment by henrygd** [comment]
@ektorasdj Awesome! You might want to unsubscribe from thread notifications while I troubleshoot with Nathan.

@nathang21 First of all, impressive resource utilization. You're definitely getting your money's worth.

The problem is indeed related to the number of containers. 24.0.2 seems to have a bug with the `one-shot` query parameter, so it can end up throttling for one second per container, which leads to a timeout if you have too many containers.

How many containers are you running in total, and do you see the same number of containers populate each time? If so, how many?

Can you try running the agent with the `LOG_LEVEL` env var set to "debug" and make sure it says `DEBUG Docker version=24.0.2 concurrency=200`?

I'm short on time tonight, but tomorrow I can give you a little bash script to verify that it's the same issue I was seeing.

**Comment by nathang21** [comment]
Sounds like progress! @henrygd Thanks for noticing :) I will say I do run BOINC which uses a limited set of spare resources available to donate compute power towards research efforts, which smooths out the CPU load around 75% or so on average (as configured). 

I have 60 containers running, and that number is fairly static (ie no dynamically scaling workloads), but is slowly trending up as I discover more things to self host in my addiction/hobby.

No rush at all, I just added the env variable, and I see the following logs which looks expected (followed by timeout spam):
```
2024/10/04 00:00:57 DEBUG Docker version=24.0.2 concurrency=200
2024/10/04 00:01:35 DEBUG Sending stats data="{Stats:{Cpu:70.18 Mem:17.42 MemUsed:6.18 MemPct:35.5 MemBuffCache:9.57 Swap:12.45 SwapUsed:4.83 DiskTotal:884.16 DiskUsed:829.07 DiskPct:93.77 DiskReadPs:25.45 DiskWritePs:111.15 NetworkSent:1.97 NetworkRecv:1.67 Temperatures:map[coretemp_core_0:63 coretemp_core_1:63 coretemp_core_2:63 coretemp_core_3:63 coretemp_physical_id_0:63] ExtraFs:map[]} Info:{Hostname:synology-nas KernelVersion:4.4.302+ Cores:4 Threads:4 CpuModel:Intel(R) Celeron(R) J4125 CPU @ 2.00GHz Uptime:547394 Cpu:70.18 MemPct:35.5 DiskPct:93.77 AgentVersion:0.5.1} Containers:[]}"
2024/10/04 00:01:35 ERROR Error getting container stats err="Get \"[http://localhost/containers/323201cdd8ac/stats?stream=0&one-shot=1\":](http://localhost/containers/323201cdd8ac/stats?stream=0&one-shot=1\%22:) context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
...
```

**Comment by henrygd** [comment]
Cool, I'm going to look into setting that up. I remember doing folding@home on my PS3 back in the day.

Here are two bash scripts which should help figure out the problem. They both send 10 requests for container stats to Docker. One does it in sequence and the other in parallel.

First run the curl command to make sure it returns the stats properly. I'm using container ID 323201cdd8ac because it was in your logs, but swap it out with something else in `docker ps` if the ID changed.

```bash
curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" "http://localhost/containers/323201cdd8ac/stats?stream=0&one-shot=1"
```

**sequence.sh**

Save this to `sequence.sh`, make it executable `chmod +x ./sequence.sh`, then run it with time: `time ./sequence.sh`

It should take 9 seconds to complete.

```bash
#!/bin/bash

for i in {1..10}
do
   curl -s --unix-socket /var/run/docker.sock -H "Content-Type: application/json" "http://localhost/containers/323201cdd8ac/stats?stream=0&one-shot=1"
done
```

**parallel.sh**

Save this to `parallel.sh`, make it executable `chmod +x ./parallel.sh`, then run it with time: `time ./parallel.sh`

It should take 1 second to complete. Also try changing 10 to 60. It should still take only 1 second.

```bash
#!/bin/bash

for i in {1..10}
do
   curl -s --unix-socket /var/run/docker.sock -H "Content-Type: application/json" "http://localhost/containers/323201cdd8ac/stats?stream=0&one-shot=1" &
done

wait
```

Please take your time, and I don't need the entire output. Just let me know if you see anything different in the timings.

Thanks!

**Comment by nathang21** [comment]
Same yeah, well feel free to reach out if you run into trouble, the linuxserver.io image is the best I think for getting it running in Docker, it emulates the old school desktop app. curl test


```
curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" "http://localhost/containers/3e47c2d6a203/stats?stream=0&one-shot=1"
{"read":"2024-10-06T03:05:50.729191617Z","preread":"0001-01-01T00:00:00Z","pids_stats":{},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":109585487,"percpu_usage":[13058467,9094015,68392415,19040590],"usage_in_kernelmode":40000000,"usage_in_usermode":20000000},"system_cpu_usage":515213000000000,"online_cpus":4,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":0,"usage_in_kernelmode":0,"usage_in_usermode":0},"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":180224,"max_usage":11902976,"stats":{"active_anon":61440,"active_file":28672,"cache":61440,"dirty":0,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":61440,"inactive_file":28672,"mapped_file":28672,"pgfault":4516,"pgmajfault":32,"pgpgin":5151,"pgpgout":5107,"rss":118784,"rss_huge":0,"total_active_anon":61440,"total_active_file":28672,"total_cache":61440,"total_dirty":0,"total_inactive_anon":61440,"total_inactive_file":28672,"total_mapped_file":28672,"total_pgfault":4516,"total_pgmajfault":32,"total_pgpgin":5151,"total_pgpgout":5107,"total_rss":118784,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":18702540800},"name":"/it-tools","id":"3e47c2d6a2030d78935c0f11785508d49732cd845715b02a03f09b5c75b2a115","networks":{"eth0":{"rx_bytes":5691558,"rx_packets":16472,"rx_errors":0,"rx_dropped":0,"tx_bytes":0,"tx_packets":0,"tx_errors":0,"tx_dropped":0}}}
```

I ran them all a few times, I was noticing some different results. Hope this lmks, lmk if you need more tests. 

sequence test
```
./sequence.sh  0.11s user 0.07s system 0% cpu 4:49.35 total
./sequence.sh  0.10s user 0.07s system 0% cpu 1:13.66 total
./sequence.sh  0.10s user 0.06s system 0% cpu 34.202 total
```

parallel test 10
```
./parallel-10.sh  0.10s user 0.06s system 23% cpu 0.667 total
./parallel-10.sh  0.09s user 0.07s system 9% cpu 1.630 total
./parallel-10.sh  0.11s user 0.12s system 2% cpu 10.193 total
```

parallel test 60:
```
./parallel-60.sh  0.53s user 0.37s system 126% cpu 0.715 total
./parallel-60.sh  0.58s user 0.33s system 6% cpu 13.480 total
./parallel-60.sh  0.56s user 0.32s system 29% cpu 2.918 total
```

**Comment by henrygd** [comment]
Thanks for doing that. Really strange results. Not sure how making 10 requests could take almost 5 minutes, or why the timings are so spread out.

My first thought is to check the health of the containers. Maybe one is stuck in a boot loop or has some other issue that is causing problems.

Try running `docker ps -a` and look for any containers that have been up less than a minute or exited very recently.

Or maybe run [ctop](https://github.com/bcicen/ctop) and look for anything irregular.

**Comment by nathang21** [comment]
Thanks for sharing ctop, really great tool I hadn't seen before. 

Indeed, I was noticing the past few days that everything was a little slow, I think my disk I/O was overloaded. I've taken significant steps to customize a bunch of my containers to prioritize using RAM instead of writing to disk, as well as reducing writes in general. 

Everything feels more stable now, however the docker graphs in Beszel are still full of gaps (more gaps than actual data), and I just re-ran the same scripts, here is the trimmed output:

```
➜  scripts time ./sequence.sh && time ./parallel-10.sh && time ./parallel-60.sh
./sequence.sh  0.11s user 0.06s system 0% cpu 18.826 total
./parallel-10.sh  0.10s user 0.07s system 8% cpu 1.874 total
./parallel-60.sh  0.64s user 0.41s system 51% cpu 2.051 total
``` 

Does this look any better?

**Comment by henrygd** [comment]
I increased the timeout to 2100ms in 0.5.3, so that's probably helping as well.

In the next release I'll use two different timeouts depending on version, and bump the older versions to eight seconds or so. That may finally fix it.

When I was testing docker 24, it was in an otherwise empty LXC container. So I was seeing the delay, but it was always consistent. In your case the inconsistency may be due to other programs also accessing the API and creating a queue. Which, again, is a bug does not happen in 25+.

**Comment by nathang21** [comment]
Just an update after the latest release, unfortunately I'm now seeing gaps in all of the graphs, not just the docker ones, which appears to be a regression.

Will share logs and more details tomorrow when I'm at my computer.

**Comment by henrygd** [comment]
I added an env var `DOCKER_TIMEOUT` in 0.6.1. You can try tweaking this to see if you can find a sweet spot where it returns data more reliably. Use values like `4s` or `3200ms`.

It looks like you're getting queue pileups that overlap time windows from Beszel requests.

I just thought of one other thing to try that I'll put in the next release.

**Comment by nathang21** [comment]
Cool thanks, I've started playing around with this but so far haven't noticed a big difference. I assume there are dimminition returns of increasing the timeout, but not quite sure how to think about it. I've tried 15s and 30s respectively but both still had gaps (now in all graphs as mentioned previously, not just the docker graphs).

**Comment by henrygd** [comment]
Let me know if you have any more success with 0.6.2. :crossed_fingers: 

I'd also recommend pausing the system for a bit in case the API queue is really jammed up.

If you want to try tweaking `DOCKER_TIMEOUT` again, your best bet is probably something between `2s` and `6s`.

**Comment by nathang21** [comment]
0.6.2 does seem better, although I paused for a bit to clear the queue, hadn't thought about that. I also reduced the timeout to `5s`, and now the regular graphs don't have gaps, but the docker graphs still have occasional ones.

Not sure which is the primary improvement (maybe a combination). It's definitely much more useable now, but not 100% resolved, this may be the limits we can reach on docker v24?


<img width="1411" alt="image" src="https://github.com/user-attachments/assets/12ea4e9f-1645-4fbc-bf20-95e5a1594976">

**Comment by henrygd** [comment]
Try dropping `DOCKER_TIMEOUT` down to `3s`  and comment out `LOG_LEVEL=debug` if you have it active.

Let me know if you still have gaps after that.

I have a couple other things to try.

**Comment by nathang21** [comment]
Done, and I did have debug logging enabled which I removed. 

That seemingly helped, but still not in the clear totally. Let me know if you have other suggestions.

Thanks for being so diligent with this.

<img width="1405" alt="image" src="https://github.com/user-attachments/assets/ac71603e-5526-4da7-8013-03b3c8e376a4">

**Comment by henrygd** [comment]
No worries, let me know if you start getting gaps in the non-docker metrics again. Or if anything else changes in a significant way.

I'll continue tweaking things on this end.

**Comment by nathang21** [comment]
Still seeing occasional gaps in non-docker metrics and docker metrics, the latter being much more common.

Overall it's definitely much more stable, and may be related to the I/O bottlenecks on my system discussed previously (which I may explore continuing to optimize).

**Comment by henrygd** [comment]
Thanks for the update. I'll have another possible fix in the next release.

I think the most likely explanation is that another service you host is also requesting info from Docker.

Every one minute, the agent tries to ask docker for the stats. Sometimes timing is bad and it hits a queue that was made by the other service.

If this is the case, it may be impossible to fully fix, but we can try to make it as stable as possible.

**Comment by henrygd** [comment]
@nathang21 I decided not to include the tweaks for old docker versions in the 0.7.0 release because I got sidetracked with localization and didn't have time to test it as much as I wanted to.

If you want to test manually, you can do this:

1. Change image name in docker compose:

```yml
# image: "henrygd/beszel-agent:latest"
image: "beszel-agent:latest
```

2. Download docker image:

```bash
wget https://henrygd-assets.b-cdn.net/beszel/bin/beszel-agent-image.zst
```

3. Load image:

```bash
docker load -i ./beszel-agent-image.zst
```

4. Run docker compose:

```bash
docker compose up -d
```

This may also be a way to test different things in a more controlled way, rather than including the changes in every new release.

**Comment by nathang21** [comment]
Sounds like good plan, thanks for all the effort on this 🙏.

FWIW Just ran the above commands, doesn't seem to be a major difference so far:
<img width="1381" alt="image" src="https://github.com/user-attachments/assets/c2a6dd0f-b0bd-4d01-b6b0-a7bd82a17816">

**Comment by mufeedali** [comment]
Is it possible to disable the warning since I use podman?

**Comment by henrygd** [comment]
I'm going to close this issue as not planned. I gave it my best shot but either there's nothing we can do or I'm too dumb to figure it out.

@mufeedali The outdated Docker version warning? Right now we check if your docker host contains "podman" to determine  if you use podman. Is your host `podman.sock` or something else?

**Comment by mufeedali** [comment]
@henrygd My bad, I was mapping the socket to `/var/run/docker.sock`. I didn't expect Beszel to include support for Podman. That's really cool! Love the software btw, thanks for building it :)
